Add support to call unpack for pytorch mobile quantized FC and Conv

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #26335 Unify Quantization APIs for add, pool and relu
* #26307 Changes to support int8 weight and fp32 bias in QNNPACK
* **#26211 Add support to call unpack for pytorch mobile quantized FC and Conv**

Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17464430](https://our.internmc.facebook.com/intern/diff/D17464430)

supriyar committed:
Add support to call unpack for pytorch mobile quantized FC and Conv
Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

dzhulgakov commented on aten/src/ATen/native/quantized/cpu/qnnpack_utils.h:
```diff
@@ -11,12 +11,16 @@ struct QnnpackOperatorDeleter {
 };
 struct PackedLinearWeightsQnnp {
   std::unique_ptr<qnnpack::PackBMatrix> w;
+  at::Tensor orig_weight;
```
I'd add a comment explaining why we do this (that implementing proper unpack function would be an alternative. Also, mention that we could easily optimize the memory by not storing the original weights in the mobile case

dzhulgakov reviewed (approved):
Looks good, but please add a comment

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422827](https://our.internmc.facebook.com/intern/diff/D17422827)

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422827](https://our.internmc.facebook.com/intern/diff/D17422827)

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422827](https://our.internmc.facebook.com/intern/diff/D17422827)

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422827](https://our.internmc.facebook.com/intern/diff/D17422827)

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422827](https://our.internmc.facebook.com/intern/diff/D17422827)

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17422827](https://our.internmc.facebook.com/intern/diff/D17422827)

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17464430](https://our.internmc.facebook.com/intern/diff/D17464430)

[ghstack-poisoned]

supriyar committed:
Update on "Add support to call unpack for pytorch mobile quantized FC and Conv"


Summary:
Currently QNNPACK does not have an unpack function like FBGEMM does.
In order to be able to script quantized models for mobile, we need to save unpacked weights.

This change stores the original weights and bias in the opaque struct and simply returns it when unpack is called

Test Plan:
python test/test_quantized.py TestQNNPackOps.test_qconv_unpack
python test/test_quantized.py TestQNNPackOps.test_qlinear_unpack

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D17464430](https://our.internmc.facebook.com/intern/diff/D17464430)

[ghstack-poisoned]
